newsfragments/3108.bugfix.rst

```diff
@@ -0,0 +1 @@
+avoid materializing frame.f_locals in KI protection code, this causes locals to persist in the frame.f_locals snapshot even if they are deleted on 3.12 and below
```

Comment:
I think this should be cast as a feature -- in addition to fixing locals materialization, I would expect it to also be good for performance and tracebacks because we're adding extra frames in far fewer places. Also, please use rST formatting and full sentences in newsfragments so our release notes are pleasant for users to read.

Reply:
could you help me out with the phrasing for this, please?

Reply:
```
Rework KeyboardInterrupt protection to track code objects, rather than frames,
as protected or not. This should substantially reduce its overhead, and also fixes
an issue with the previous implementation: locals' lifetimes will no longer be
extended by materialization in the ``frame.f_locals`` dictionary that the previous
KI protection logic needed to access to do its work.
```

Reply:
```suggestion
Rework KeyboardInterrupt protection to track code objects, rather than frames,
as protected or not. This should substantially reduce its overhead, and also fixes
an issue with the previous implementation: locals' lifetimes will no longer be
extended by materialization in the ``frame.f_locals`` dictionary that the previous
KI protection logic needed to access to do its work.
```